rust/catalyst-types/src/uuid/mod.rs

```diff
@@ -32,16 +33,30 @@ pub enum CborUuidError {
     },
 }
 
-/// Decode `CBOR` encoded `UUID`.
-pub(crate) fn decode_cbor_uuid(val: &coset::cbor::Value) -> Result<uuid::Uuid, CborUuidError> {
-    let Some((UUID_CBOR_TAG, coset::cbor::Value::Bytes(bytes))) = val.as_tag() else {
-        return Err(CborUuidError::InvalidCborType);
-    };
-    let uuid = uuid::Uuid::from_bytes(
-        bytes
-            .clone()
-            .try_into()
-            .map_err(|_| CborUuidError::InvalidByteSize)?,
-    );
+/// Decode from `CBOR` into `UUID`
+pub(crate) fn decode_cbor_uuid(
+    d: &mut Decoder<'_>, (): &mut (),
+) -> Result<uuid::Uuid, minicbor::decode::Error> {
+    let tag = d.tag()?;
+    if UUID_CBOR_TAG != tag.as_u64() {
+        return Err(minicbor::decode::Error::message(format!(
+            "tag value must be: {UUID_CBOR_TAG}, provided: {}",
+            tag.as_u64(),
+        )));
+    }
+    let decoded = d
+        .bytes()?
+        .try_into()
+        .map_err(|_| minicbor::decode::Error::message("Expected UUID to have 16 bytes"))?;
+    let uuid = uuid::Uuid::from_bytes(decoded);
     Ok(uuid)
 }
+
+/// Encode `UUID` into `CBOR`
+pub(crate) fn encode_cbor_uuid<W: minicbor::encode::Write>(
```

Comment:
For generality, encoding with a Tag should be optional.
Suggest adding a context `tag: bool`.
* true = encode with a tag
* false = encode without a tag.